casanovo/denovo/model.py

```diff
@@ -176,12 +193,592 @@ def forward(
+        cache_seq_dict: Dict[int, Set[str]],
+        cache_score_dict: Dict[int, List[List[Tuple[float, int]]]],
+        cache_tokens: torch.Tensor,
+        cache_scores: torch.Tensor,
+        tokens: torch.Tensor,
+        precursors: torch.Tensor,
+        scores: torch.Tensor,
+        idx: int,
+    ) -> Tuple[
+        Dict[int, int],
+        Dict[int, Set[str]],
+        Dict[int, List[List[Tuple[float, int]]]],
+    ]:
+        """
+        Cache terminated beams, group and order by fitting precursor m/z and confidence score
+
+        Parameters
+        ----------
+        finished_beams_idx : List[int]
+            Indices of all finished beams on tokens tensor
+        cache_idx_dict : Dict[int, int]
+            Next available tensor index to cache peptides for each spectrum.
+        cache_seq_dict : Dict[int,  Set[str]]
+            Set of decoded peptides for each spectrum.
+        cache_score_dict : Dict[int, List[List[Tuple[float, int]]]
+            Confidence score for each decoded peptide, separated as precursor m/z fitting vs not, for each spectrum.
+        cache_tokens : torch.Tensor of shape
+        (n_spectra* n_beams, max_length)
+            The token for each amino acid in cached peptides.
+        cache_scores : torch.Tensor of shape
+        (n_spectra* n_beams, max_length, n_amino_acids)
+            The score for each amino acid in cached peptides.
+        tokens : torch.Tensor of shape (n_spectra* n_beams, max_length)
+            Output token of the model corresponding to amino acid sequences of peptides.
+        precursors : torch.Tensor of size (n_spectra* n_beams, 3)
+            The measured precursor mass (axis 0), precursor charge (axis 1), and
+            precursor m/z (axis 2) of each MS/MS spectrum.
+         scores : torch.Tensor of shape (n_spectra*  n_beams, max_length, n_amino_acids)
+            Output scores of the model.
+        idx : int
+            Index to be considered in the current decoding step
+
+        Returns
+        -------
+        finished_beams_idx : List[int]
+            Indices of all finished beams on tokens tensor
+        cache_idx_dict : Dict[int, int]
+            Next available tensor index to cache peptides for each spectrum.
+        cache_seq_dict : Dict[int,  Set[str]]
+            Set of decoded peptides for each spectrum.
+        cache_score_dict : Dict[int, List[List[Tuple[float, int]]]
+            Confidence score for each decoded peptide, separated as precursor m/z fitting vs not, for each spectrum.
+
+        """
+        beam = self.n_beams
+        # Store finished beams in the cache
+        if len(finished_beams_idx) > 0:
+            for i in finished_beams_idx:
+                spec_idx = i // beam  # find the starting index of the spectrum
+                insert_idx = cache_idx_dict[spec_idx]
+
+                # Only add beams with matching mass to the cache
+                prec_mz = precursors[i, 2].item()
+                aa_list = [
```

Comment:
**question:** We're decoding the AAs and calculating the mass both when terminating and caching beams. Can we avoid doing this twice?

Reply:
Will cache delta m/z to avoid duplication but couldn't come up with a straight forward way of caching the peptide to avoid idx2AA decoding twice i.e. beams can terminate when stop token predicted before precursor m/z is exceeded which causes all sorts of edge cases